### ISSUE TITLE
Allow controlling of Invocation lifetime/scope

### ISSUE THREAD

**Issue by BishoyDemian** [Azure/azure-webjobs-sdk#716]
I'm using the SDK with Autofac dependency injection, and a very important scenario is to control lifetime scope of dependency injection and disposal when invoking a new function.

I used the `IJobActivator` interface to plug-in my Autofac container, however I cannot activate my types within a lifetime scope because there are no callbacks to dispose of the lifetime when the function invocation is finished (succeeded or not).

What I'm doing now is this:

```
        public T CreateInstance<T>()
        {
            var scope = _container.BeginLifetimeScope();
            return scope.Resolve<T>();
        }
```

The code above is bad because the lifetimescope is `IDisposable`, and will not be disposed automatically for me, and there is no way for me to determine when the activated instance has been disposed in order to dispose my lifetime scope.

I propose adding an `ActivationContext` parameter that is passed per activation (ideally invocation), that has the function invocation unique id. Then adding a `Deactivate()` method that takes the same context, to allow me to dispose of the lifetime I created earlier.

**Comment by mathewc** [comment]
Can't you already get the "deactivate" behavior you're looking for by making your job class `IDisposable`? When we create an instance of your job class via `IJobActivator`, after the invocation is complete, if it is `IDisposable` we will dispose the instance. That would allow you to dispose whatever you need.

**Comment by BishoyDemian** [comment]
Well, in theory, yes this alternative solution might work, however when the `IJobActivator` resolves the function class it should also resolve and inject multiple levels of dependencies that have several other lifetime/scope configurations (singletons, per invocation, per thread, etc.).

I personally don't like my business function (or its class) to worry about or (even worse) control the scope/lifetime of its dependencies directly, as this goes against SRP. And my container already comes with lifetime management facility that I can use.

I'd be happy to undertake the work required here in a fork and submit a PR, however I wanted to kick start the discussion here first to gather as much information and feedback as possible before I start refactoring any code?

**Comment by mathewc** [comment]
It sounds like the changes you're proposing would be breaking changes on IJobActivator, so they require careful thought. You're basically looking for a callback for when a job class instance is disposed. IDisposable is meant for exactly that, so it seems really natural to me. Perhaps there is a way to use that while also maintaining your DI design purity? :)

**Comment by abatishchev** [comment]
Is disposing IDisposable already implemented or only is a proposal?

**Comment by BishoyDemian** [comment]
@mathewc If changing `IJobActivator` is not desirable, then how about making other interfaces that perform execution (and disposal) public?

**Comment by jamesbascle** [comment]
Is there some kind of event or state change, that the framework emits that indicates a job class is done with that we could hook into in the job activator?  Ultimately, one needs to be able to have their container manage their lifetimes, rather than just treat it as a strict service locator.  Because of that, the current IJobActivator seems like something of a half measure.

**Comment by abatishchev** [comment]
Yeah, it definitely needs improvement to play nicer with common DI container contract (such as accept Type, return Object)

**Comment by jamesbascle** [comment]
In the meantime, I was seemingly able to workaround this deficiency by:
1)  giving all of my job classes a base class (I've called mine "Reactor") with just a 
````C#
public abstract void React(string messagePath)
````
signature. (obviously change this to suit your own architectural needs)

2) Intercepting it using Castle
3) Having the interceptor take in the relevant lifetime scope in its constructor.
4) Checking that the correct method is being intercepted, and disposing the lifetime scope after letting the interceptor call the method.

For reference, I'm using the Autofac IoC container, and the Autofac.Extra.DynamicProxy2 NuGet package, with latest Castle.Core.

Here's the code that shows how to set things up. https://gist.github.com/jamesbascle/d02deb717a9e8baa0c47eaaf6a41230c

Just need to create a new lifetime scope without a using block before the call to Container.Resolve in your IJobActivator.

**Comment by zsims** [comment]
A simpler approach (at least for autofac) is to take a `ILifetimeScope` into the job, create a child scope, r

**Comment by abatishchev** [comment]
What I'm looking for is a generic (DI container agnostic) solution without injecting DI container specific components (since obvisok that may vary) and without heavy lifting such as proxies and interceptions.

**Comment by jamesbascle** [comment]
IMO, the interception overhead isn't a particularly big deal - it happens
the once at the composition root, and it doesn't require injecting the
scope to use the container as a service locator.  Obviously a legit IoC
hook would be ideal, but it'll do for me.

**Comment by drub0y** [comment]
So this seems to have stagnated, so I just wanted to chime in and say that I started down this path and found `IJobActivator` and was like 😀, but then quickly realized there was no ability to control scope and was like 😢.

I spent a little time looking at the SDK current design in hopes of offering some advice on how to maybe offer scoping, but unfortunately I'm not quite sure how to get out of this corner it's current painted into right now. If backwards compatibility is a concern then the best option may actually be leaving `IJobActivator` the way it is, maybe even obsoleting it at some point, and instead revisiting the solution to DI from the top and offering it as a sibling option to the job activator approach. Then the upstream runtime logic could basically check if this "new" DI approach was configured and, if so, it would just ignore any existing activator concepts (maybe even throw and exception if both are used) and prefer the new DI approach instead.

**Comment by nagoh** [comment]
This implementation goes someway to controlling scope in a non-invasive manner? http://stackoverflow.com/a/33759649/2111224

**Comment by abatishchev** [comment]
I agree it should be implemented/provided by the framework on the lowest level.

**Comment by emparker** [comment]
I would like this functionality as well. At the moment. I have to create container scope inside my function and wrapped it in using. It is working but it would be nice to have framework handle only worry about business login inside my function :-)
